nameresolution/consul/watcher.go

```diff
@@ -0,0 +1,342 @@
+		r.logger.Errorf("consul service-watcher error: %v, retry in %v", err, retry)
+		sleepTimer := time.NewTimer(retry)
+		select {
+		case <-ctx.Done():
+			sleepTimer.Stop()
+			r.logger.Debugf("consul service-watcher retry throttling canceled")
+		case <-sleepTimer.C:
+		}
+
+		return
+	}
+
+	// reset the plan failure count
+	p.failures = 0
+
+	// if the index is unchanged do nothing
+	if p.lastParamVal != nil && p.lastParamVal.equal(blockParam) {
+		return
+	}
+
+	// update the plan index
+	oldParamVal := p.lastParamVal
+	p.lastParamVal = blockParam.next(oldParamVal)
+
+	// compare last and new result to get changed services
+	healthByService := getHealthByService(result)
+	changedServices := p.getChangedServices(healthByService)
+
+	// update the plan last result
+	p.lastResult = healthByService
+
+	// call agent to get updated healthy nodes for each changed service
+	for k := range changedServices {
+		p.options.WaitIndex = 0
+		p.options.Filter = p.configuredQueryFilter
+		p.options = p.options.WithContext(ctx)
+		result, _, err := r.client.Health().Service(k, "", true, p.options)
+
+		if err != nil {
+			// on failure, expire service from cache, resolver will fall back to agent
+			r.logger.Errorf("error invoking health service: %v, for service %s", err, k)
+			r.registry.expire(k)
+
+			// remove healthchecks for service from last result
+			for key := range p.lastResult {
+				if k == key.serviceName {
+					delete(p.lastResult, key)
+				}
+			}
+
+			// reset plan query index
+			p.lastParamVal = waitIndexVal(0)
+		} else {
+			// updated service entries in registry
+			r.logger.Debugf("updating registry for service:%s last-index:%s", k, strconv.FormatUint(uint64(p.lastParamVal.(waitIndexVal)), 10))
+			r.registry.addOrUpdate(k, result)
+		}
+	}
+}
+
+func (r *resolver) runWatchLoop(p *watchPlan) {
+	defer func() {
+		recover()
+		r.registry.removeAll()
+		r.watcherStarted = false
+	}()
+
+	watchTask := make(chan bool, 1)
+
+	for {
+		ctx, cancel := context.WithCancel(context.Background())
+
+		services := r.registry.getKeys()
+		watching := false
+
+		if len(services) > 0 {
+			go r.runWatchPlan(p, services, ctx, watchTask)
+			watching = true
+		}
+
+		select {
+		case <-watchTask:
+			cancel()
+
+		// wait on channel for new services to track
+		case service := <-r.registry.registrationChannel():
+			// cancel blocking query to consul agent
+			cancel()
+
+			// generate set of keys
+			serviceKeys := make(map[string]interface{})
+			for i := 0; i < len(services); i++ {
+				serviceKeys[services[i]] = nil
+			}
+
+			// add service if it's not in the registry
+			if _, ok := serviceKeys[service]; !ok {
+				r.registry.addOrUpdate(service, nil)
+			}
+
+			// check for any more new services in channel
+			moreServices := true
+			for moreServices {
+				select {
+				case service := <-r.registry.registrationChannel():
+					if _, ok := serviceKeys[service]; !ok {
+						r.registry.addOrUpdate(service, nil)
+					}
+				default:
```

Comment:
The problem with this approach is that you are racing with the producer. If the producer adds to the channel faster than you consume them, you will end the loop sooner.

Consider maybe making `registrationChannel` accept a slice so you can send all keys to add in bulk, in a single message over the channel?

Reply:
Sorry, there's a mistake in this comment - I've rephrased both possibilities of the produce/consume statement and addressed below.

> If the producer adds to the channel faster than you consume them, **_the loop will not end_**.

This is only possible if the user is constantly resolving new IDs, in that bizarre case caching would be irrelevant.

> If the **_consumer reads from_** the channel faster than **_the producer writes_**, you will end the loop sooner.

This is obviously the expected behaviour.

The registration channel is only written to if a new service/app-id that is not in the registry is being resolved - there is really no batching concept here so it doesn't quite make sense to make it a slice channel. To elaborate:

#### ResolveID (sidecar invoke)
1. Resolver checks registry for key
2. Key does not exist and is written to channel
3. Resolver invokes consul agent directly

#### Watcher routine (perpetual)
1. Watcher picks up key from the channel and cancels long poll
2. Watcher adds the key to the registry
3. Watcher checks for any other new keys on channel
4. Watcher invokes long poll with new set of keys